### PR TITLE
LF: Move KV's InputsAndEffects to LF land

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
We move the code from computing from KV to LF. 
Testing + more refactoring will follows in subsequent PRs.  

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
@andreaslochbihler-da , @simonmaxen-da Do the function `computeInputs` and `computeEffects' are usefull for you ?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala [hunk 1]
```diff
@@ -100,4 +104,112 @@ object Util {
...
 
+  /** Compute the inputs to a DAML transaction (in traversal order), that is, the referenced contracts, parties,
+    *  and key.
+    */
+  def collectInputs(
+      tx: Transaction.Transaction,
+      addContract: ContractId => Unit,
+      addParty: Ref.Party => Unit,
+      addGlobalKey: GlobalKey => Unit,
+  ): Unit = {
```
Comment (SamirTalwar-DA) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala:115]: I'm not a big fan of `Util` classes/objects because they tend to become a dumping ground for any random function.

I would prefer if this function lived in its own object, as before.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Effects.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+/** @param consumedContracts
+  *     The contracts consumed by this transaction.
+  *     When committing the transaction these contracts must be marked consumed.
+  *     A contract should be marked consumed when the transaction is committed,
+  *     regardless of the ledger effective time of the transaction (e.g. a transaction
+  *     with an earlier ledger effective time that gets committed later would find the
+  *     contract inactive).
+  * @param createdContracts
+  *     The contracts created by this transaction.
+  *     When the transaction is committed, keys marking the activeness of these
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Effects.scala:20]: `keys` sounds very much like we're targeting a KV store here.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+package com.daml.lf.transaction
+
+import com.daml.lf.data.Ref
+import com.daml.lf.value.Value.ContractId
+
+import scala.collection.immutable.TreeSet
+
+/** The inputs of the transaction, that is:
+  *   - the contracts were fetched
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala:12]: What about contracts that were returned by a `lookupByKey`? And what about contracts that are exercised, but not fetched?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  *   - all the parties involved
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala:13]: What does "involved" mean? The parties that are an informee of some node in the transaction? Or all party IDs mentioned in the transaction? Or something else?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  *   - all the contract keys involved.
+  */
+/** @param contracts
+  *     The contracts fetched by this transaction.
+  *
+  * @param parties
+  *     The contracts created by this transaction.
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala:20]: This line doesn't fit with the `parties` parameter name. This comment looks like a copy-n-paste error.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  *     When the transaction is committed, keys marking the activeness of these
+  *     contracts should be created. The key should be a combination of the transaction
+  *     id and the relative contract id (that is, the node index).
+  * @param keys
+  *     The contract keys created or updated as part of the transaction.
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala:26]: This information is not enough to check the preconditions of a transaction w.r.t. the ledger state. For example, it's not clear whether the keys should be unassigned or assigned at the beginning of the transaction.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: done.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala [hunk 7]
```diff
@@ -100,4 +104,112 @@ object Util {
...
+
+    val localContract = tx.localContracts
+
+    def addContractInput(coid: ContractId): Unit =
+      if (!localContract.isDefinedAt(coid))
+        addContract(coid)
+
+    def addPartyInputs(parties: Set[Ref.Party]): Unit =
+      parties.toList.sorted(Ref.Party.ordering).foreach(addParty)
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: Why do you have to sort the parties?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala [hunk 8]
```diff
@@ -100,4 +104,112 @@ object Util {
...
+        case fetch: Node.NodeFetch[Value.ContractId] =>
+          addContractInput(fetch.coid)
+          fetch.key.foreach(addContractKey(fetch.templateId, _))
+
+        case create: Node.NodeCreate[Value.ContractId] =>
+          create.key.foreach(addContractKey(create.templateId, _))
+
+        case exe: Node.NodeExercises[NodeId, Value.ContractId] =>
+          addContractInput(exe.targetCoid)
+          exe.key.foreach(addContractKey(exe.templateId, _))
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: Why do you have to add the key to the inputs if the `byKey` flag is not set?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: KV needs it to get deterministic results.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: We don't. If the `byKey` flag is not set, then `exe.key = None`.

**Review by andreaslochbihler-da — COMMENTED** [review]
I don't think these functions are particularly useful for Canton as is.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: Thanks!

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala (hunk 6 above)
Comment (simonmaxen-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala:26]: Yes, I need this to.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala:26]: andreaslochbihler-da
> This information is not enough to check the preconditions of a transaction w.r.t. the ledger state. For example, it's not clear > whether the keys should be unassigned or assigned at the beginning of the transaction.

Let's tackle that in an upcoming PR.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: I drop the sort.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala]: Let's postpone semantics consideration to another PR. 
For now, I just move the code around keeping as much as possible the original behavior of the utility as defined in KV.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+case class Inputs(
+    contracts: TreeSet[ContractId],
+    parties: TreeSet[Ref.Party],
+    keys: TreeSet[GlobalKey],
+)
+
+object Inputs {
+
+  /** Compute the inputs to a DAML transaction (in traversal order), that is, the referenced contracts, parties,
+    *  and key.
```
Comment (miklos-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Inputs.scala:30]:
```suggestion
  /** Compute the inputs to a Daml transaction (in traversal order), that is, the referenced contracts, parties,
    *  and keys.
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala [hunk 10]
```diff
@@ -57,19 +64,21 @@ private[state] object Conversions {
...
-      )
+    assertDecode("Indentifier", ValueCoder.decodeIdentifier(protoIdent))
 
   def globalKeyToStateKey(key: GlobalKey): DamlStateKey = {
     DamlStateKey.newBuilder
       .setContractKey(encodeGlobalKey(key))
       .build
   }
 
+  def contractKeyToStateKey(tmplId: Identifier, key: Value[ContractId]): DamlStateKey =
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:75]:
```suggestion
  def contractKeyToStateKey(templateId: Identifier, key: Value[ContractId]): DamlStateKey =
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Effects.scala [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.transaction
+
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value.ContractId
+
+import scala.collection.immutable.{TreeMap, TreeSet}
+
+/** The effects of a transaction, that is:
```
Comment (miklos-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Effects.scala:11]: Are the comments in the original `InputsAndEffects.Effects` regarding semantics of consumed/create contracts captured somewhere else in the code?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala [hunk 12]
```diff
@@ -44,14 +49,26 @@ class KeyValueSubmission(metrics: Metrics) {
...
-      val inputDamlStateFromTx = InputsAndEffects.computeInputs(tx, meta)
+      val packageDamlStates = meta.optUsedPackages
+        .getOrElse(
+          throw new InternalError("Transaction was not annotated with used packages")
+        )
+        .map(Conversions.packageStateKey)
+      val inputDamlStatesFromTx = new util.ArrayList[DamlStateKey]()
+      transaction.Inputs.collectInputs(
+        tx,
+        { cid => inputDamlStatesFromTx.add(Conversions.contractIdToStateKey(cid)); () },
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala:60]: Why not use native Scala types here for improving readability and calling `.asJava` when constructing the protobuf message?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala [hunk 13]
```diff
@@ -1,7 +1,8 @@
 // Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-package com.daml.ledger.participant.state
+package com.daml
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala:4]: Please revert this change. We don't use this package notation in `kvutils`.

**Review by miklos-da — COMMENTED** [review]
Looks good overall, thanks, @remyhaemmerle-da!

**Comment by remyhaemmerle-da** [comment]
subsumed by  #9429 and  #9433

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala (hunk 13 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala:4]: I revert this change in  #9433

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala (hunk 12 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala:60]: I apply the suggestion in  #9429

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala (hunk 10 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:75]: I applied this in  #9429 and #9433
